analysis/src/code.rs

```diff
@@ -0,0 +1,152 @@
+    pub is_direct: bool,
+    pub loc_report: LOCReport,
+    pub dep_report: DepReport,
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct LOCReport {
+    pub is_analyzed: bool,
+    pub total_code_loc: u64, // excludes comment and white lines
+    pub rust_code_loc: u64,  // excludes comment and white lines
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, Clone)]
+pub struct DepReport {
+    pub is_analyzed: bool,
+    pub total_deps: u64,
+    pub deps_total_code_loc: u64,
+    pub deps_rust_code_loc: u64,
+}
+
+pub struct CodeAnalyzer {
+    cache: RefCell<HashMap<String, LOCReport>>,
+}
+
+impl CodeAnalyzer {
+    pub fn new() -> Self {
+        Self {
+            cache: RefCell::new(HashMap::new()),
+        }
+    }
+
+    pub fn analyze_code(self, graph: &PackageGraph) -> Result<Vec<CodeReport>> {
+        let mut code_reports: Vec<CodeReport> = Vec::new();
+
+        // Get direct dependencies
+        let direct_dependencies: Vec<PackageMetadata> = graph
+            .query_workspace()
+            .resolve_with_fn(|_, link| {
+                let (from, to) = link.endpoints();
+                from.in_workspace() && !to.in_workspace()
+            })
+            .packages(guppy::graph::DependencyDirection::Forward)
+            .filter(|pkg| !pkg.in_workspace())
+            .collect();
+
+        for package in &direct_dependencies {
+            let loc_report = self.get_loc_report(package.manifest_path())?;
+            let dependencies: Vec<PackageMetadata> = graph
+                .query_forward(iter::once(package.id()))?
+                .resolve()
+                .packages(DependencyDirection::Forward)
+                .filter(|pkg| pkg.id() != package.id())
+                .collect();
+            let dep_report = self.get_dep_report(dependencies)?;
+
+            let code_report = CodeReport {
+                name: package.name().to_string(),
+                version: package.version().to_string(),
+                is_direct: true,
+                loc_report,
+                dep_report,
+            };
+
+            code_reports.push(code_report);
+        }
+
+        Ok(code_reports)
+    }
+
+    fn get_dep_report(&self, dependencies: Vec<PackageMetadata>) -> Result<DepReport> {
+        let total_deps = dependencies.len() as u64;
+        let mut deps_total_code_loc = 0;
+        let mut deps_rust_code_loc = 0;
+
+        for package in &dependencies {
+            let loc_report = self.get_loc_report(package.manifest_path())?;
+            deps_total_code_loc += loc_report.total_code_loc;
+            deps_rust_code_loc += loc_report.rust_code_loc;
+        }
+        Ok(DepReport {
+            is_analyzed: true,
+            total_deps,
+            deps_total_code_loc,
+            deps_rust_code_loc,
+        })
+    }
+
+    fn get_loc_report(&self, manifest_path: &Utf8Path) -> Result<LOCReport> {
+        if self.cache.borrow().contains_key(&manifest_path.to_string()) {
```

Comment:
`Utf8Path` impls `AsRef<str>` so you should be able to to `as_ref()` here instead `to_string()` which will save the unncessary allocation.

Reply:
my cache is `cache: RefCell<HashMap<String, LOCReport>>`, that'd mean I'd need to to `&manifest_path.as_str().to_string()` anyways, or am I missing something?